File: Backend/flower_store/util/builder/order_builder.py
from typing import List

from flower_store.model.bouquet import Bouquet
from flower_store.model.order import Order


class OrderBuilder:
    def __init__(self):
        self._price = None
        self._deliveryType = None
        self._clientId = None
        self._bouquets = []

    def set_price(self, price: int):
        self._price = price
        return self

    def set_deliveryType(self, deliveryType: str):
        self._deliveryType = deliveryType
        return self

    def set_clientId(self, clientId: int):
        self._clientId = clientId
        return self

    def add_bouquet(self, bouquet: Bouquet):
        self._bouquets.append(bouquet)
        return self

    def build(self):
        if self._price is None or self._deliveryType is None or self._clientId is None:
            raise ValueError("Price, DeliveryType, and ClientId must be set")
        return Order(
            price=self._price,
            deliveryType=self._deliveryType,
            clientId=self._clientId,
            bouquets=self._bouquets
        )

    def prepareOrder(self, price: int, delivery_type: str, client_id: int, bouquets: List[Bouquet]) -> Order:
        order = Order(
            price=price,
            deliveryType=delivery_type,
            clientId=client_id
        )

        if bouquets:
            for bouquet in bouquets:
                order.addItemToOrder(bouquet)

        return order

File: Frontend/src/pattern/OrderBuilder.js
class Order {
    constructor(price, deliveryType, clientId, bouquets) {
        this.price = price;
        this.deliveryType = deliveryType;
        this.clientId = clientId;
        this.bouquets = bouquets;
    }
}

export class OrderBuilder {
    constructor() {
        this.reset();
    }

    reset() {
        this.price = 0;
        this.deliveryType = '';
        this.clientId = 0;
        this.bouquets = [];
    }

    setPrice(price) {
        this.price = price;
        return this;
    }

    setDeliveryType(deliveryType) {
        this.deliveryType = deliveryType;
        return this;
    }

    setClientId(clientId) {
        this.clientId = clientId;
        return this;
    }

    setBouquets(bouquetData) {
        bouquetData.forEach(bouquetItem => {
            const { item, count } = bouquetItem;
            for (let i = 0; i < count; i++) {
                this.bouquets.push(item);
            }
        });
        return this;
    }

    build() {
        const order = new Order(this.price, this.deliveryType, this.clientId, this.bouquets);
        this.reset();
        return order;
    }
}